### ISSUE TITLE
[feature] add affine parameterization, synchronized commitment loss or alternating optimization

### ISSUE THREAD

**Issue by kashif** [lucidrains/vector-quantize-pytorch#47]
from the paper:

[Straightening Out the Straight-Through Estimator: Overcoming Optimization Challenges in Vector Quantized Networks, Huh et al. ICML2023](https://minyoungg.github.io/vqtorch/)

**Comment by kashif** [comment]
all good! i hope you are well too! I made this issue more for myself to try out and report. I will continue to read and then try out and let you know!

**Comment by lucidrains** [comment]
@kashif yes, do let me know, it looks like a very simple intervention!

**Comment by lucidrains** [comment]
hmm actually on second look, they had to combine a few interventions. also it seems their sync v value ranges from 0.01 to 2. that alone makes it unattractive

let me think about reinmax some more and see if it can be adapted here (and run a few experiments when i get a stretch of time)

**Comment by lucidrains** [comment]
<img width="416" alt="Screen Shot 2023-05-24 at 10 56 51 AM" src="https://github.com/lucidrains/vector-quantize-pytorch/assets/108653/0ba48b9d-5e23-4f2a-9af9-5a269c3d6170">

hey, good news! i think i'm seeing something after adding straight-through with reinmax

would you like to give it a try and see if it does anything for your problem?

**Comment by kashif** [comment]
ah nice! ok let me try too!

**Comment by lucidrains** [comment]
haha, actually disregard my anecdata for now. the effect only seems to be there for cosine sim

i'll have to set aside a morning for a full set of experiments

**Comment by lucidrains** [comment]
also decided to try the stochastic code sampling on a whim and seeing much better earlier convergence (only works for euclidean dist)

**Comment by kashif** [comment]
going over the reinmax, I wanted to share this taxonomy of gradient estimators with respect to the Straight-through estimator which falls under the biased-low-variance type:

![gradient-estimators drawio](https://github.com/lucidrains/vector-quantize-pytorch/assets/8100/1658bb1c-c8c0-413a-b8f6-2a85562fc9f8)

the image is from a presentation from Paulus et al, 2021, and I thought you might find it useful

**Comment by lucidrains** [comment]
@kashif embarrassingly enough, there's a bug in the straight through implementation, will fix it this morning

**Comment by lucidrains** [comment]
https://wandb.ai/lucidrains/rvq-vae-gpt-4/reports/testing-vq-lib--Vmlldzo0NDY4MTQy

running a few experiments before moving onto some vision transformers work

**Comment by lucidrains** [comment]
yea, i'm not seeing it for reinmax. i think i'll have to read this new MIT paper in depth and try out some of their conclusions

**Comment by lucidrains** [comment]
@kashif hey Kashif, i read the paper in more depth today and i'm sold on the affine parameterization

will try to get that built (and make sure it can handle the batch statistics in a distributed setting) by next week's end

**Comment by lucidrains** [comment]
@kashif ok, verified that it runs at the very least, and no worse than baseline

whether it actually alleviates index collapse i'll leave up crowdsourcing

will get the distributed version done at some point

re: alternating optimization - i think that can be done external to the framework? that is simply where one learns on the commit loss vs the actual task loss?

**Comment by lucidrains** [comment]
@kashif affine parameterization should work distributed now, and i've also added that new hyperparameter that controls the synchronous update

**Comment by kashif** [comment]
thank you! i will test it out!

**Comment by lucidrains** [comment]
@kashif so i think in the paper they are not using exponential moving averaged updates, so you'll have to also pass in

```python
VectorQuantize(
   learnable_codebook = True,
   ema_update = False
)
```

should probably just make the codebook default learnable if the EMA update is turned off

**Comment by lucidrains** [comment]
i think i will probably do a big cleanup of the repository, only keeping what works by next month's end. there's too many conflicting features

**Comment by Froskekongen** [comment]
> hmm, difference is really slight, at least, for my architecture
> 
> stochastic sampling leads to a dramatic difference in early convergence though. count me surprised

What about early code utilization?  I got the impression when reading the straight-through paper that code collapse often happens early in training.  Struggling with this code collapse a bit with a particularly difficult dataset.

**Comment by theAdamColton** [comment]
I'm not sure if I'm setting up the vq using a learnable codebook correctly. In this small example the codebook nn.Parameter embed has no grad

```python
vq = VectorQuantize(dim=512,
                    codebook_size=32,
                    codebook_dim=512,
                    heads=1,
                    ema_update=False,
                    learnable_codebook=True,
                )
x = torch.rand(7, 1, 512)
*_, loss = vq(x)
loss.backward()
print(vq._codebook.embed.grad) # this is None
```

**Comment by kashif** [comment]
@theAdamColton  its a bug i believe here:

https://github.com/lucidrains/vector-quantize-pytorch/blob/master/vector_quantize_pytorch/vector_quantize_pytorch.py#L408

this should be:

```
embed = self.embed if self.learnable_codebook else self.embed.detach()
```

**Comment by lucidrains** [comment]
@kashif thanks Kashif

to do open source is to have all your embarrassing mistakes put on record 😅

**Comment by kashif** [comment]
indeed! no shame in open source!

**Comment by theAdamColton** [comment]
Learnable codebook still doesn't work after this [commit](https://github.com/lucidrains/vector-quantize-pytorch/commit/d23d1fd730ef3a129459ba4ae1463607f5e4d108), you'll still have no embedding gradients if you use the codebook like this:

```python
vq = VectorQuantize(dim=512,
                    codebook_size=32,
                    codebook_dim=512,
                    heads=1,
                    ema_update=False,
                    learnable_codebook=True,
                )
x = torch.rand(7, 1, 512)
*_, loss = vq(x)
loss.backward()
print(vq._codebook.embed.grad) # this is None
```
This is due to the straight-through approximation [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/ae5db0169772ae4c624e65d3b5c33f45cdd34a92/vector_quantize_pytorch/vector_quantize_pytorch.py#LL801C15-L801C15) which detaches the quantized vector from the nn.Parameter learnable embedding. Any optimization of the `vq._codebook.embed` will have to be done before doing straight-through, I think. 

In the vq-ste++ code for in-place/alternating optimization they pass an optimizer to the VQ constructor, and update the codebook embedding [before doing straight-through](https://github.com/minyoungg/vqtorch/blob/406021e6d67c607aa987baf1e79a1de90d2bba3c/vqtorch/nn/vq.py#L120-L129). 

It might not be completely out of the scope of this project to pass an inplace_optimizer to the VQ constructor

**Comment by lucidrains** [comment]
@theAdamColton ah yea i'll have to take a look at that once i get back from my trip.

for now, i put in a quick fix for the sync update

**Comment by lucidrains** [comment]
> Learnable codebook still doesn't work after this [commit](https://github.com/lucidrains/vector-quantize-pytorch/commit/d23d1fd730ef3a129459ba4ae1463607f5e4d108), you'll still have no embedding gradients if you use the codebook like this:
> 
> ```python
> vq = VectorQuantize(dim=512,
>                     codebook_size=32,
>                     codebook_dim=512,
>                     heads=1,
>                     ema_update=False,
>                     learnable_codebook=True,
>                 )
> x = torch.rand(7, 1, 512)
> *_, loss = vq(x)
> loss.backward()
> print(vq._codebook.embed.grad) # this is None
> ```
> 
> This is due to the straight-through approximation [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/ae5db0169772ae4c624e65d3b5c33f45cdd34a92/vector_quantize_pytorch/vector_quantize_pytorch.py#LL801C15-L801C15) which detaches the quantized vector from the nn.Parameter learnable embedding. Any optimization of the `vq._codebook.embed` will have to be done before doing straight-through, I think.
> 
> In the vq-ste++ code for in-place/alternating optimization they pass an optimizer to the VQ constructor, and update the codebook embedding [before doing straight-through](https://github.com/minyoungg/vqtorch/blob/406021e6d67c607aa987baf1e79a1de90d2bba3c/vqtorch/nn/vq.py#L120-L129).
> 
> It might not be completely out of the scope of this project to pass an inplace_optimizer to the VQ constructor

this should be addressed

other than the alternating optimization, i think the other two features should be there

closing
